[SPARK-24923][SQL][WIP] Add unpartitioned CTAS and RTAS support for DataSourceV2

rdblue opened the pull request:
## What changes were proposed in this pull request?

* Remove extends from `ReadSupport` and `WriteSupport` classes for use with `Table`
* Add CTAS and RTAS logical plans
* Refactor physical write plans so AppendData, CTAS, and RTAS use the same base class
* Add support for `TableCatalog` to `DataFrameReader` and `DataFrameWriter`
* Add `TableV2Relation` for tables that are loaded by `TableCatalog` and have no `DataSource` instance
* Move implicit helpers into `DataSourceV2Implicits` to avoid future churn

Note that this doesn't handle `partitionBy` in `DataFrameWriter`. Adding support for partitioned tables will require validation rules.

This is based on unmerged work and includes the commits from #21306 and #21305.

## How was this patch tested?

Adding unit tests for CTAS and RTAS.

rdblue committed:
SPARK-24252: Add v2 data source mix-in for catalog support.

rdblue committed:
SPARK-24923: Add v2 CTAS and RTAS support.

This uses the catalog API introduced in SPARK-24252 to implement CTAS
and RTAS plans.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1330/
Test PASSed.

SparkQA commented:
**[Test build #93572 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93572/testReport)** for PR 21877 at commit [`d308d3c`](https://github.com/apache/spark/commit/d308d3c75f78242c822eab6d11fb651d94f10aa6).

rdblue commented:
@cloud-fan, @gatorsmile, @marmbrus, this PR demonstrates how plans would use the catalog changes introduced in #21306. To see the changes, you may want to look at just the last commit because this includes changes from other PRs.

SparkQA commented:
**[Test build #93572 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93572/testReport)** for PR 21877 at commit [`d308d3c`](https://github.com/apache/spark/commit/d308d3c75f78242c822eab6d11fb651d94f10aa6).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class CreateTableAsSelect(`
  * `case class ReplaceTableAsSelect(`
  * `case class TableV2Relation(`
  * `case class AppendDataExec(`
  * `case class CreateTableAsSelectExec(`
  * `case class ReplaceTableAsSelectExec(`
  * `case class WriteToDataSourceV2Exec(`
  * `abstract case class V2TableWriteExec(`
  * `  implicit class CatalogHelper(catalog: CatalogProvider) `
  * `  implicit class TableHelper(table: Table) `
  * `  implicit class SourceHelper(source: DataSourceV2) `
  * `  implicit class OptionsHelper(options: Map[String, String]) `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93572/
Test FAILed.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1365/
Test PASSed.

SparkQA commented:
**[Test build #93613 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93613/testReport)** for PR 21877 at commit [`5dcf159`](https://github.com/apache/spark/commit/5dcf159f0d2a2cf599a3e4e22e0fbd0a6f283bf1).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1367/
Test PASSed.

SparkQA commented:
**[Test build #93616 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93616/testReport)** for PR 21877 at commit [`8709957`](https://github.com/apache/spark/commit/87099572303f3539ec935fde89f6a044ecaa093c).

SparkQA commented:
**[Test build #93615 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93615/testReport)** for PR 21877 at commit [`48c9998`](https://github.com/apache/spark/commit/48c99986343ce8419cd0cce40e926506b1ee37f4).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93615/
Test FAILed.

SparkQA commented:
**[Test build #93616 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93616/testReport)** for PR 21877 at commit [`8709957`](https://github.com/apache/spark/commit/87099572303f3539ec935fde89f6a044ecaa093c).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class CreateTableAsSelect(`
  * `case class ReplaceTableAsSelect(`
  * `case class TableV2Relation(`
  * `case class AppendDataExec(`
  * `case class CreateTableAsSelectExec(`
  * `case class ReplaceTableAsSelectExec(`
  * `case class WriteToDataSourceV2Exec(`
  * `abstract case class V2TableWriteExec(`
  * `  implicit class CatalogHelper(catalog: CatalogProvider) `
  * `  implicit class TableHelper(table: Table) `
  * `  implicit class SourceHelper(source: DataSourceV2) `
  * `  implicit class OptionsHelper(options: Map[String, String]) `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93616/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1369/
Test PASSed.

SparkQA commented:
**[Test build #93618 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93618/testReport)** for PR 21877 at commit [`65e42b9`](https://github.com/apache/spark/commit/65e42b99168fb2d148f4d042bbc18b97e08df7f7).

SparkQA commented:
**[Test build #93618 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93618/testReport)** for PR 21877 at commit [`65e42b9`](https://github.com/apache/spark/commit/65e42b99168fb2d148f4d042bbc18b97e08df7f7).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class CreateTableAsSelect(`
  * `case class ReplaceTableAsSelect(`
  * `case class TableV2Relation(`
  * `case class AppendDataExec(`
  * `case class CreateTableAsSelectExec(`
  * `case class ReplaceTableAsSelectExec(`
  * `case class WriteToDataSourceV2Exec(`
  * `abstract case class V2TableWriteExec(`
  * `  implicit class CatalogHelper(catalog: CatalogProvider) `
  * `  implicit class TableHelper(table: Table) `
  * `  implicit class SourceHelper(source: DataSourceV2) `
  * `  implicit class OptionsHelper(options: Map[String, String]) `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93618/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1370/
Test PASSed.

SparkQA commented:
**[Test build #93620 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93620/testReport)** for PR 21877 at commit [`37b981b`](https://github.com/apache/spark/commit/37b981b88f0b99c0e73fc84ed3ef6a5c50c627fd).

SparkQA commented:
**[Test build #93620 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93620/testReport)** for PR 21877 at commit [`37b981b`](https://github.com/apache/spark/commit/37b981b88f0b99c0e73fc84ed3ef6a5c50c627fd).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class CreateTableAsSelect(`
  * `case class ReplaceTableAsSelect(`
  * `case class TableV2Relation(`
  * `case class AppendDataExec(`
  * `case class CreateTableAsSelectExec(`
  * `case class ReplaceTableAsSelectExec(`
  * `case class WriteToDataSourceV2Exec(`
  * `abstract class V2TableWriteExec(`
  * `  implicit class CatalogHelper(catalog: CatalogProvider) `
  * `  implicit class TableHelper(table: Table) `
  * `  implicit class SourceHelper(source: DataSourceV2) `
  * `  implicit class OptionsHelper(options: Map[String, String]) `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93620/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/1379/
Test PASSed.

SparkQA commented:
**[Test build #93638 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93638/testReport)** for PR 21877 at commit [`b6b29d8`](https://github.com/apache/spark/commit/b6b29d809bb76ea59b7ec66e8a80b224a938495b).

SparkQA commented:
**[Test build #93638 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/93638/testReport)** for PR 21877 at commit [`b6b29d8`](https://github.com/apache/spark/commit/b6b29d809bb76ea59b7ec66e8a80b224a938495b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class CreateTableAsSelect(`
  * `case class ReplaceTableAsSelect(`
  * `case class TableV2Relation(`
  * `case class AppendDataExec(`
  * `case class CreateTableAsSelectExec(`
  * `case class ReplaceTableAsSelectExec(`
  * `case class WriteToDataSourceV2Exec(`
  * `abstract class V2TableWriteExec(`
  * `  implicit class CatalogHelper(catalog: CatalogProvider) `
  * `  implicit class TableHelper(table: Table) `
  * `  implicit class SourceHelper(source: DataSourceV2) `
  * `  implicit class OptionsHelper(options: Map[String, String]) `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/93638/
Test PASSed.

rdblue committed:
SPARK-24252: Add PartitionTransform to replace Expression.

Expression is internal and should not be used in public APIs. To avoid
using Expression in the TableCatalog API, this commit adds a small set
of transformations that are used to communicate partitioning to catalog
implementations.

This also adds an apply transformation that passes the name of a
transform instead of a Transform class. This can be used to pass
transforms that are unknown to Spark to the underlying catalog
implementation.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/2229/
Test PASSed.

SparkQA commented:
**[Test build #94828 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94828/testReport)** for PR 21877 at commit [`e50d94b`](https://github.com/apache/spark/commit/e50d94bbc205369969e0c1707bda057ee99f0007).

SparkQA commented:
**[Test build #94828 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/94828/testReport)** for PR 21877 at commit [`e50d94b`](https://github.com/apache/spark/commit/e50d94bbc205369969e0c1707bda057ee99f0007).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class CreateTableAsSelect(`
  * `case class ReplaceTableAsSelect(`
  * `case class TableV2Relation(`
  * `case class AppendDataExec(`
  * `case class CreateTableAsSelectExec(`
  * `case class ReplaceTableAsSelectExec(`
  * `case class WriteToDataSourceV2Exec(`
  * `abstract class V2TableWriteExec(`
  * `  implicit class CatalogHelper(catalog: CatalogProvider) `
  * `  implicit class TableHelper(table: Table) `
  * `  implicit class SourceHelper(source: DataSourceV2) `
  * `  implicit class OptionsHelper(options: Map[String, String]) `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/94828/
Test FAILed.

SparkQA commented:
**[Test build #99893 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99893/testReport)** for PR 21877 at commit [`e50d94b`](https://github.com/apache/spark/commit/e50d94bbc205369969e0c1707bda057ee99f0007).

SparkQA commented:
**[Test build #99893 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99893/testReport)** for PR 21877 at commit [`e50d94b`](https://github.com/apache/spark/commit/e50d94bbc205369969e0c1707bda057ee99f0007).
 * This patch **fails to build**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `case class CreateTableAsSelect(`
  * `case class ReplaceTableAsSelect(`
  * `case class TableV2Relation(`
  * `case class AppendDataExec(`
  * `case class CreateTableAsSelectExec(`
  * `case class ReplaceTableAsSelectExec(`
  * `case class WriteToDataSourceV2Exec(`
  * `abstract class V2TableWriteExec(`
  * `  implicit class CatalogHelper(catalog: CatalogProvider) `
  * `  implicit class TableHelper(table: Table) `
  * `  implicit class SourceHelper(source: DataSourceV2) `
  * `  implicit class OptionsHelper(options: Map[String, String]) `

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99893/
Test FAILed.

SparkQA commented:
**[Test build #101100 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101100/testReport)** for PR 21877 at commit [`e50d94b`](https://github.com/apache/spark/commit/e50d94bbc205369969e0c1707bda057ee99f0007).

SparkQA commented:
**[Test build #101100 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101100/testReport)** for PR 21877 at commit [`e50d94b`](https://github.com/apache/spark/commit/e50d94bbc205369969e0c1707bda057ee99f0007).
 * This patch **fails to build**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `case class CreateTableAsSelect(`
  * `case class ReplaceTableAsSelect(`
  * `case class TableV2Relation(`
  * `case class AppendDataExec(`
  * `case class CreateTableAsSelectExec(`
  * `case class ReplaceTableAsSelectExec(`
  * `case class WriteToDataSourceV2Exec(`
  * `abstract class V2TableWriteExec(`
  * `  implicit class CatalogHelper(catalog: CatalogProvider) `
  * `  implicit class TableHelper(table: Table) `
  * `  implicit class SourceHelper(source: DataSourceV2) `
  * `  implicit class OptionsHelper(options: Map[String, String]) `

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101100/
Test FAILed.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/15753/
Test PASSed.
